File: Java/Graphs/ProblemsOnDfsBfs/801-is-graph-bipartite/is-graph-bipartite.java
class Solution {
    private boolean dfs(int node,int currentColor,  int adjList[][], int color[] ){
      color[node] = currentColor;
        
        for(int neighbour : adjList[node]){

              if(color[neighbour] == -1){
                  if(dfs(neighbour, 1-currentColor, adjList,color) == false){
                    return false;
                  }
              }else if(color[neighbour] == color[node]){
                return false;
              }
        }
      return true;
    }
    public boolean isBipartite(int[][] graph) {
        int color[] = new int[graph.length];
        for(int i=0;i<color.length;i++)color[i] = -1;

        for(int i=0;i<color.length;i++){
          if(color[i] == -1){
            if(dfs(i,0, graph,color) == false){
              return false;
            }
          }
        }

        return true;
    }
}

File: Python/Graphs/ProblemsOnBFSAndDFS/801-is-graph-bipartite/is-graph-bipartite.py
class Solution:
    def bfs(self, start:int,adjList: List[List[int]], color:List[int]):
        q = deque()
        q.append(start)

        color[start] = 0

        while q:
            node = q.popleft()

            for neighbour in adjList[node]:
                if color[neighbour] == -1:
                    color[neighbour] = 1- color[node]
                    q.append(neighbour)
                elif color[neighbour] == color[node]:
                    return False

        return True


    def isBipartite(self, graph: List[List[int]]) -> bool:
        color : List[int] = [-1] * len(graph)

        for i in range(len(color)):
          if(color[i] == -1):
            if(self.bfs(i, graph,color) == False):
              return False
            
        

        return True